libraries/Microsoft.Bot.Builder.Azure/CosmosDbSqlStorage.cs

```diff
@@ -0,0 +1,215 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.Documents;
+using Microsoft.Azure.Documents.Client;
+using Microsoft.Azure.Documents.Linq;
+using Microsoft.Bot.Builder.Core.Extensions;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Azure
+{
+    /// <summary>
+    /// Models IStorage around a CosmosDB SQL (DocumentDB)
+    /// </summary>
+    public class CosmosDbSqlStorage : IStorage
+    {
+        private readonly string databaseId;
+        private readonly string collectionId;
+        private readonly DocumentClient client;
+
+        private const string UserAgentSuffix = "Microsoft-BotFramework 4.0.0";
+
+        private static JsonSerializer jsonSerializer = JsonSerializer.Create(new JsonSerializerSettings()
+        {
+            TypeNameHandling = TypeNameHandling.All
+        });
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CosmosDbSqlStorage"/> class,
+        /// using the provided CosmosDB credentials, DatabaseId and CollectionId.
+        /// </summary>
+        /// <param name="serviceEndpoint">The endpoint Uri for the service endpoint from the Azure Cosmos DB service.</param>
+        /// <param name="authKey">The AuthKey used by the client from the Azure Cosmos DB service.</param>
+        /// <param name="databaseId">The Database ID.</param>
+        /// <param name="collectionId">The Collection ID.</param>
+        public CosmosDbSqlStorage(Uri serviceEndpoint, string authKey, string databaseId, string collectionId, Action<ConnectionPolicy> connectionPolicyConfigurator = null)
+        {
+            if (serviceEndpoint == null)
+            {
+                throw new ArgumentNullException(nameof(serviceEndpoint), "Service EndPoint for CosmosDB is required.");
+            }
+
+            if (string.IsNullOrEmpty(authKey))
+            {
+                throw new ArgumentException("AuthKey for CosmosDB is required.", nameof(authKey));
+            }
+
+            if (string.IsNullOrEmpty(databaseId))
+            {
+                throw new ArgumentException("DatabaseId is required.", nameof(databaseId));
+            }
+
+            if (string.IsNullOrEmpty(collectionId))
+            {
+                throw new ArgumentException("CollectionId is required.", nameof(collectionId));
+            }
+
+            this.databaseId = databaseId;
+            this.collectionId = collectionId;
+
+            var connectionPolicy = new ConnectionPolicy()
+            {
+                UserAgentSuffix = UserAgentSuffix
+            };
+
+            connectionPolicyConfigurator?.Invoke(connectionPolicy);
+
+            this.client = new DocumentClient(serviceEndpoint, authKey, connectionPolicy);
+            this.client.CreateDatabaseIfNotExistsAsync(new Database { Id = databaseId })
```

Comment:
I'm a bit concerened with doing this in the constructor. My concern stems from this typically being created from inside the ASP.Net DI framework, which makes debugging exceptions from constructors fairly difficult. 

Any other options you can think of?

Reply:
All operations could check for the correct initialization through a Lazy value.
The NodeJS counter-part is using this approach: [initialization](https://github.com/Microsoft/botbuilder-js/pull/165/files#diff-e0d3beb663e06b054d55b37493892559R164) & [checking initialization](https://github.com/Microsoft/botbuilder-js/pull/165/files#diff-e0d3beb663e06b054d55b37493892559R86).